### ISSUE TITLE
Pester Test Runner Fails with Script hashtable parameter

### ISSUE THREAD

**Issue by tobitux** [rfennell/vNextBuild#358]
### VSTS Extension you are using
Pester Test Runner Build Task

### Where are you running it?
VSTS with on-premise agent:

$PSVersionTable output:
PSVersion                      4.0
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   6.3.9600.18968
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}
PSRemotingProtocolVersion      2.2

### Version of Extension/Task
 8.4.2

### Expected behaviour and actual behaviour
**Expected:** 
Pester tests are executed 

**Actual:**
Task fails:
2018-08-16T16:47:07.8735971Z ##[section]Starting: Pester Test Runner
2018-08-16T16:47:07.8743473Z ==============================================================================
2018-08-16T16:47:07.8743961Z Task         : Pester Test Runner from Black Marble
2018-08-16T16:47:07.8744494Z Description  : Run Pester tests by either installing the latest version of Pester at run time (if possible) or using the version shipped with the task (4.3.1)
2018-08-16T16:47:07.8745009Z Version      : 8.4.2
2018-08-16T16:47:07.8745325Z Author       : Richard Fennell
2018-08-16T16:47:07.8745778Z Help         : Version: 8.4.2. [More Information](https://github.com/rfennell/vNextBuild/wiki/Pester-Task/)
2018-08-16T16:47:07.8746267Z ==============================================================================
2018-08-16T16:47:07.8940342Z Preparing task execution handler.
2018-08-16T16:47:08.2475379Z Executing the powershell script: C:\vstsagent\A1\_work\_tasks\Pester_31ef0033-64e3-4c55-b888-f446541474a6\8.4.2\Pester.ps1
2018-08-16T16:47:08.4036592Z Running in AMD64 PowerShell
2018-08-16T16:47:08.4037157Z 
2018-08-16T16:47:08.4037338Z 
2018-08-16T16:47:08.4059512Z Adding additional module path [] to $env:PSModulePath
2018-08-16T16:47:08.4060124Z 
2018-08-16T16:47:08.4060405Z 
2018-08-16T16:47:10.1929347Z Running Pester from using the script parameter [@{Path='C:\vstsagent\A1\_work\r2\a'; Parameters=@{param1='111'; param2='222'}}"] output sent to [C:\vstsagent\A1\_work\r2\a\Test-Pester.XML]
2018-08-16T16:47:10.1929821Z 
2018-08-16T16:47:10.1929935Z 
2018-08-16T16:47:10.2216232Z ##[error]System.Management.Automation.RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not contain a method named 'SafeGetValue'.
2018-08-16T16:47:10.2230097Z    at CallSite.Target(Closure , CallSite , Object )
2018-08-16T16:47:10.2230516Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
2018-08-16T16:47:10.2230955Z    at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
2018-08-16T16:47:10.2231417Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-08-16T16:47:10.3081752Z Executing all tests in ''
2018-08-16T16:47:10.3082085Z 
2018-08-16T16:47:10.3082228Z 
2018-08-16T16:47:10.3498653Z Tests completed in 0ms
2018-08-16T16:47:10.3499007Z 
2018-08-16T16:47:10.3499150Z 
2018-08-16T16:47:10.3638922Z Tests Passed: 0, 
2018-08-16T16:47:10.3654431Z Failed: 0, 
2018-08-16T16:47:10.3670151Z Skipped: 0, 
2018-08-16T16:47:10.3685859Z Pending: 0, 
2018-08-16T16:47:10.3702278Z Inconclusive: 0 
2018-08-16T16:47:10.3702544Z 
2018-08-16T16:47:10.3702746Z 
2018-08-16T16:47:10.5975021Z ##[error]PowerShell script completed with 1 errors.
2018-08-16T16:47:10.5987508Z ##[section]Finishing: Pester Test Runner

### Steps to reproduce the problem
1. Setup Release Definition with task
2. Set parameter "Scripts Folder or Script" to @{Path='$(System.DefaultWorkingDirectory)'; Parameters=@{param1='111'; param2='222'}}
3. Set parameter Results File to $(System.DefaultWorkingDirectory)\Test-Pester.XML
4. Run Release

![image](https://user-images.githubusercontent.com/3409712/44222772-ed5fc900-a185-11e8-8194-61433a1f5d8c.png)

**Comment by tobitux** [comment]
It happens also with PSv5.1
```
C:\> $PSVersionTable
Name                           Value
PSVersion                      5.1.17134.165
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17134.165
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

The error seems to be caused by the following line in HelperModule.psm1:
   ```
[System.Management.Automation.Language.Parser]::ParseInput($line, [Ref]$null, [Ref]$null).Find({
            $args[0] -is [System.Management.Automation.Language.HashtableAst]
        }, $false).SafeGetValue()
    
```

Maybe this helps.

**Comment by ChrisLGardner** [comment]
I'm having trouble replicating this behaviour. Can you try cloning this repo and then running Invoke-Pester on this file: https://github.com/rfennell/vNextBuild/blob/master/Extensions/Pester/test/PesterTask/Helper.Tests.ps1

That should have all the various test cases we're currently using including the example you're provided (roughly). Do you still get the same behaviour that way? 

Looking up the docs for the type we're using and [it definitely has that method](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.language.hashtableast?view=powershellsdk-1.1.0)

**Comment by tobitux** [comment]
The tests failed.

```
Type: PS D:\git\vNextBuild\Extensions\Pester\test\PesterTask> Invoke-Pester .\Helper.Tests.ps1
Executing all tests in '.\Helper.Tests.ps1'

Executing script .\Helper.Tests.ps1

  Describing Testing Helper Functions

    Context Testing Get-HashtableFromString
      [+] Can parse empty block 277ms
      [+] Can parse block with no values but delimiter 18ms
      [+] Cannot parse block with invalid string 19ms
      [-] Can parse two part block 10ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 28
      [-] Can parse two hashtable part block 17ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 35
      [-] Can parse block with trailing ; 13ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 43
      [-] Can parse three part block 1 12ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 50
      [-] Can parse three part block 2 12ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 58
      [-] Can parse hashtable with ; in a value 13ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 65
      [-] Can parse hashtable with commas in a value 14ms
        RuntimeException: Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not
 contain a method named 'SafeGetValue'.
        at Get-HashtableFromString, D:\git\vNextBuild\Extensions\Pester\Task\HelperModule.psm1: line 15
        at <ScriptBlock>, D:\git\vNextBuild\Extensions\Pester\test\PesterTask\Helper.Tests.ps1: line 73
Tests completed in 410ms
Tests Passed: 3, Failed: 7, Skipped: 0, Pending: 0, Inconclusive: 0
```

The type should have the method. Its inherited from Ast.

I tried the separate the issue for tests:
```
$line = "@{Path=''; Parameters=@{Content=''; BaseUri=''}}"
$table = [System.Management.Automation.Language.Parser]::ParseInput($line, [Ref]$null, [Ref]$null).Find({$args[0] -is [System.Management.Automation.Language.HashtableAst]}, $false);
$table.GetType()
$table | Get-Member
$table.SafeGetValue()
```

the output is: 

```
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     HashtableAst                             System.Management.Automation.Language.ExpressionAst

TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Copy
MemberType : Method
Definition : System.Management.Automation.Language.Ast Copy()


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Equals
MemberType : Method
Definition : bool Equals(System.Object obj)


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Find
MemberType : Method
Definition : System.Management.Automation.Language.Ast
             Find(System.Func[System.Management.Automation.Language.Ast,bool] predicate, bool searchNestedScriptBlocks)


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : FindAll
MemberType : Method
Definition : System.Collections.Generic.IEnumerable[System.Management.Automation.Language.Ast]
             FindAll(System.Func[System.Management.Automation.Language.Ast,bool] predicate, bool
             searchNestedScriptBlocks)


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : GetHashCode
MemberType : Method
Definition : int GetHashCode()


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : GetType
MemberType : Method
Definition : type GetType()


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : ToString
MemberType : Method
Definition : string ToString()


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Visit
MemberType : Method
Definition : System.Object Visit(System.Management.Automation.Language.ICustomAstVisitor astVisitor), void
             Visit(System.Management.Automation.Language.AstVisitor astVisitor)


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Extent
MemberType : Property
Definition : System.Management.Automation.Language.IScriptExtent Extent {get;}


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : KeyValuePairs
MemberType : Property
Definition : System.Collections.ObjectModel.ReadOnlyCollection[System.Tuple[System.Management.Automation.Language.Expre
             ssionAst,System.Management.Automation.Language.StatementAst]] KeyValuePairs {get;}


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : Parent
MemberType : Property
Definition : System.Management.Automation.Language.Ast Parent {get;}


TypeName   : System.Management.Automation.Language.HashtableAst
Name       : StaticType
MemberType : Property
Definition : type StaticType {get;}

Method invocation failed because [System.Management.Automation.Language.HashtableAst] does not contain a method named
'SafeGetValue'.
At C:\vstsagent\A1\_work\_tasks\Pester_31ef0033-64e3-4c55-b888-f446541474a6\8.4.2\test.ps1:5 char:1
+ $table.SafeGetValue()
+ ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound

```

The method is missing?

**Comment by tobitux** [comment]
Update: Tests run well at PSv5
```
Name                           Value
----                           -----
PSVersion                      5.1.17134.165
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17134.165
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1


```And failed at PSv4:

```
Name                           Value
----                           -----
PSVersion                      4.0
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   6.3.9600.18968
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}
PSRemotingProtocolVersion      2.2

```

**Comment by ChrisLGardner** [comment]
I've got a work around for this that I'll look into implementing later today. It's a little cludgy but I'll put in a warning or two about how it works and why it's happening on v4 and lower versions of PS.

**Comment by ChrisLGardner** [comment]
Released as v8.5.1

@tobitux can you give it a try ton your PSv4 machine and see if that fixes your issues.

**Comment by ChrisLGardner** [comment]
Closing this as it seems to be resolved. 

@tobitux if there are still issues then feel free to reopen or create a new issue.
